google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -0,0 +1,353 @@
+    // the database root, if it is a root collection).
+    ResourcePath parentPath = collectionReference.getResourcePath();
+    parentPath = parentPath.popLast();
+    String collectionId = collectionReference.getId();
+
+    Query query = baseAllDescendantsQuery(parentPath, collectionId);
+
+    // To find all descendants of a collection reference, we need to use a
+    // composite filter that captures all documents that start with the
+    // collection prefix. The MIN_ID constant represents the minimum key in
+    // this collection, and a null byte + the MIN_ID represents the minimum
+    // key is the next possible collection.
+    char nullChar = '\0';
+    String startAt = collectionId + "/" + REFERENCE_NAME_MIN_ID;
+    String endAt = collectionId + nullChar + "/" + REFERENCE_NAME_MIN_ID;
+    query =
+        query
+            .whereGreaterThanOrEqualTo(FieldPath.documentId(), startAt)
+            .whereLessThan(FieldPath.documentId(), endAt);
+
+    // startAfter() needs to be added after the where() filters since it creates an implicit
+    // orderBy.
+    if (lastDocumentSnap != null) {
+      query = query.startAfter(lastDocumentSnap);
+    }
+
+    return query;
+  }
+
+  private Query getAllDescendantsQuery(@Nonnull DocumentReference documentReference) {
+    // The parent is the closest ancestor document to the location we're deleting. Since we are
+    // deleting a document, the parent is the path of that document.
+    ResourcePath parentPath = documentReference.getResourcePath();
+    String collectionId = documentReference.getParent().getId();
+
+    return baseAllDescendantsQuery(parentPath, collectionId);
+  }
+
+  private Query baseAllDescendantsQuery(ResourcePath resourcePath, String collectionId) {
+    Query query =
+        new Query(
+            firestoreRpcContext,
+            QueryOptions.builder()
+                .setParentPath(resourcePath)
+                .setCollectionId(collectionId)
+                .setAllDescendants(true)
+                .setKindless(true)
+                .setRequireConsistency(false)
+                .build());
+
+    // Query for names only to fetch empty snapshots.
+    query = query.select(FieldPath.documentId()).limit(maxPendingOps);
+
+    return query;
+  }
+
+  /**
+   * Called when all descendants of the provided reference have been streamed or if a permanent
+   * error occurs during the stream. Deletes the developer provided reference and wraps any errors
+   * that occurred.
+   */
+  private void onQueryEnd() {
+    documentsPending = false;
+
+    // Used to aggregate flushFuture and deleteFuture to use with ApiFutures.allAsList(), in order
+    // to ensure that the delete catchingAsync() callback is run before the flushFuture callback.
+    List<ApiFuture<Void>> pendingFutures = new ArrayList<>();
+
+    // Delete the provided document reference if one was provided.
+    if (documentReference != null) {
+      ApiFuture<Void> voidDeleteFuture =
```

Comment:
Can we call `deleteReference`? (Not sure)

Reply:
You can make it return an ApiFuture.

```
private ApiFuture<Void> deleteReference(final DocumentReference reference) {
    synchronized (lock) {
      pendingOperationsCount++;
    }
    
    return ApiFutures.transformAsync(
          catchingDelete(reference),
          new ApiAsyncFunction<WriteResult, Void>() {

            public ApiFuture<Void> apply(WriteResult result) {
              synchronized (lock) {
                pendingOperationsCount--;
                // We wait until the previous stream has ended in order to ensure the
                // startAfter document is correct. Starting the next stream while
                // there are pending operations allows Firestore to maximize
                // BulkWriter throughput.
                if (documentsPending
                    && !streamInProgress
                    && pendingOperationsCount < minPendingOps) {
                  streamDescendants();
                }
                return ApiFutures.immediateFuture(null);
              }
            }
          },
          MoreExecutors.directExecutor());
    }
  }
```

Reply:
Ah, I see what you mean now, thanks. Done.